File: frontend/src/app/api/manualSetup.ts
// Manual assessment setup entry — for a property with no DRE/CC&R
// extraction run on file. Creates a synthetic extraction run that the
// operator then reviews and approves through the existing DRE/CC&R
// Review Workbench, exactly like any Gemini-derived run.
//
//   POST /hoa/{hoa_id}/assessment-setup/manual → ManualExtractionRunResponse

import { BASE_URL } from './config';
import { authHeaders, handleResponse } from './http';

export type PromptSetupType =
  | 'fixed_equal'
  | 'grouped_category'
  | 'individual_unit'
  | 'multi_pool_combination'
  | 'unknown_needs_review';

export type PromptAllocationMethod =
  | 'equal'
  | 'square_footage'
  | 'ownership_percentage'
  | 'category'
  | 'specified_value'
  | 'parking_space'
  | 'custom_factor'
  | 'unknown';

export interface ManualPoolEntry {
  pool_key: string;
  pool_name?: string;
  annual_amount?: number | null;
  allocation_method: PromptAllocationMethod;
  recipient_scope?: string;
  denominator_value?: number | null;
  variable_flag?: boolean;
}

export interface ManualGroupEntry {
  group_id?: string;
  label?: string;
  unit_count: number;
  average_square_feet?: number | null;
  ownership_percent?: number | null;
}

export interface ManualUnitEntry {
  unit_number: string;
  square_feet?: number | null;
  ownership_percent?: number | null;
  category?: string;
  parking_spaces?: number;
}

export interface ManualExtractionRunResponse {
  dre_document_id: number;
  extraction_run_id: number;
  property_id: number;
}

export async function createManualAssessmentSetup(
  hoaId: number,
  body: {
    setup_type: PromptSetupType;
    pools: ManualPoolEntry[];
    groups?: ManualGroupEntry[];
    units?: ManualUnitEntry[];
  },
): Promise<ManualExtractionRunResponse> {
  const res = await fetch(`${BASE_URL}/hoa/${hoaId}/assessment-setup/manual`, {
    method: 'POST',
    headers: { ...authHeaders(), 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return handleResponse(res);
}

File: backend/app/routers/manual_setup.py
"""Manual assessment setup entry endpoint — for a property with no
DRE/CC&R extraction run on file.

Creates a synthetic ``dre_extraction_runs`` row from operator-entered
pools/groups/units; the caller then approves it through the existing
DRE or CC&R approve endpoints exactly like any other run (see
``manual_assessment_setup_service`` for why there's no separate
promotion path).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from ..ai_implementation.db import get_session
from ..auth.dependencies import get_current_user
from ..dre_extraction.schemas import PromptSetupType
from ..services.manual_assessment_setup_service import (
    ManualExtractionRunResponse,
    ManualGroupEntry,
    ManualPoolEntry,
    ManualUnitEntry,
    PropertyNotFound,
    create_manual_extraction_run,
)


router = APIRouter(tags=["Manual Assessment Setup"])


def _actor_email(actor: dict) -> str:
    return str(actor.get("email") or actor.get("name") or "unknown")


class ManualSetupEntryRequest(BaseModel):
    setup_type: PromptSetupType
    pools: list[ManualPoolEntry]
    groups: list[ManualGroupEntry] = []
    units: list[ManualUnitEntry] = []


@router.post(
    "/hoa/{hoa_id}/assessment-setup/manual",
    response_model=ManualExtractionRunResponse,
    status_code=201,
)
def create_manual_setup_entry(
    hoa_id: int,
    payload: ManualSetupEntryRequest,
    session: Session = Depends(get_session),
    current_user: dict = Depends(get_current_user),
) -> ManualExtractionRunResponse:
    """Create a manually-entered extraction run for a property with no
    DRE/CC&R document on file.

    Returns the new ``extraction_run_id``; the operator reviews it in the
    same Review Workbench and approves it via the existing DRE/CC&R
    approve endpoints (use the CC&R endpoint for a ``per_unit`` setup to
    get the proportional-pool missing-factor guard).
    """
    raw_conn = session.connection().connection
    try:
        return create_manual_extraction_run(
            property_id=hoa_id,
            setup_type=payload.setup_type,
            pools=payload.pools,
            groups=payload.groups,
            units=payload.units,
            created_by=_actor_email(current_user),
            connection=raw_conn,
        )
    except PropertyNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
